Allow ending specific span

hauleth opened the pull request:
This will allow for finer graded control which span ends when.

hauleth committed:
feat(tracer): allow ending specific span

This will allow for finer graded control which span ends when.

hauleth committed:
feat(tracer): remove dangerous with_trace/2 from public API

hauleth committed:
feat(tracer): move more logic to the API

This reduces amount of thing the implementor needs to think about and
makes whole API more bullet-proof. Additionally rename `with_span/2` to
`set_span/2` to reduce confusion between behaviour of these two.

tsloughter commented on include/opentelemetry.hrl:
```diff
@@ -50,7 +50,9 @@
           %% trace flags lowest bit is 1 but simply not propagated.
-          is_recorded       :: boolean() | undefined
+          is_recorded       :: boolean() | undefined,
+          %% parent span
```
Nice, i was going to replace the 2-tuple with a new record `trace_ctx` that held the current span and the parent `trace_ctx` but this is better. It just needs a more detailed comment here to make clear why this is in the span context.

tsloughter commented on src/ot_tracer.erl:
```diff
@@ -31,27 +30,26 @@
-with_span(Tracer={Module, _}, SpanCtx, Fun) when is_atom(Module) ->
-    Module:with_value(Tracer, SpanCtx, Fun).
+    SpanCtx = Module:start_span(Tracer, Name, Opts),
+    Module:set_span(Tracer, SpanCtx),
```
Stuff like this shouldn't be in the API. It is up to the SDK.

hauleth replied:
I tend to disagree. Mostly because this is a thing that **all** tracer modules will need to do anyway. There is no way to escape this. For me moving it single layer above makes job simpler for implementors who will not need to think about setting proper variables on their own. It will simply reduce the cognitive overhead and will make tracer module much simpler. I would even go and rename callback functions `start_span` -> `create_span` and `end_span` -> `store_span`.

tsloughter replied:
Na, it isn't, at least not in the current spec. It has been discussed, and the API is not meant to do this, an SDK could never set a span to the context.

hauleth replied:
If I [read this](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#span-creation) correctly then neither should SDK set it as an active span. We only should expose `set_span/2` function as well.

Strange decision to make IMHO, but ok.

tsloughter commented on src/ot_tracer.erl:
```diff
@@ -31,27 +30,26 @@
+-spec end_span(opentelemetry:tracer(), opentelemetry:span_ctx()) -> ok.
+end_span(Tracer={Module, _}, SpanCtx) ->
+    Module:end_span(Tracer, SpanCtx),
+    Module:set_span(Tracer, SpanCtx#span_ctx.parent).
```
Same as above.

tsloughter commented on src/ot_tracer.erl:
```diff
@@ -31,27 +30,26 @@
-with_span(Tracer={Module, _}, Span) when is_atom(Module) ->
-    Module:with_span(Tracer, Span).
-
--spec with_span(opentelemetry:tracer(), opentelemetry:span_ctx(), fun()) -> ok.
```
`ot_tracer` needs a `with_span` that takes a function. Otherwise a user can't provide both a `Tracer` and a `fun`.

hauleth replied:
I do not get what do you mean. Obviously they will not be able as such behaviour was moved one layer further to the `otel` module to simplify interface.

hauleth committed:
feat(otel): change the with_span/3 function callback

Now it gets span context as an argument on which all context actions
should happen. If user want to then they can propagate it themselves.
